src/metabase/query_processor/parameters.clj

```diff
@@ -0,0 +1,228 @@
+        (substitute-all-params query rest)))
+    query-dict))
+
+(def ^:private ^:const outer-clause #"\[\[.*?\]\]")
+(def ^:private ^:const outer-clause-prefix #"^\[\[(.*?)\s.*\]\]$")
+(def ^:private ^:const incomplete-outer-clause #"\[\[.*?\{\{.*?\}\}.*?\]\]")
+(def ^:private ^:const inner-clause #"<(.*?)>")
+(def ^:private ^:const incomplete-inner-clause #"<.*?\{\{.*?\}\}.*?>")
+
+(defn- remove-incomplete-clauses [query-dict]
+  (let [find-and-replace (fn [sql]
+                           (-> sql
+                               (s/replace incomplete-outer-clause "")
+                               (s/replace incomplete-inner-clause "")))]
+    (update-in query-dict [:native :query] find-and-replace)))
+
+(defn- conjoin-multi-clause [clause]
+  (let [prefix (second (re-find outer-clause-prefix clause))]
+    ;; re-seq produces a vector for each match like [matched-form grouping1] and we only want grouping1.
+    (str prefix " " (s/join " AND " (map second (re-seq inner-clause clause))))))
+
+(defn- process-multi-clauses [query-dict]
+  (if-let [multi-clauses (re-seq outer-clause (get-in query-dict [:native :query]))]
+    (update-in query-dict [:native :query] (fn [q]
+                                             (loop [sql                   q
+                                                    [multi-clause & rest] multi-clauses]
+                                               (if multi-clause
+                                                 (recur (s/replace-first sql multi-clause (conjoin-multi-clause multi-clause)) rest)
+                                                 sql))))
+    query-dict))
+
+(defn- process-single-clauses [query-dict]
+  (if-let [single-clauses (re-seq inner-clause (get-in query-dict [:native :query]))]
+    (update-in query-dict [:native :query] (fn [q]
+                                             (loop [sql                      q
+                                                    [[orig stripped] & rest] single-clauses]
+                                               (if orig
+                                                 (recur (s/replace-first sql orig stripped) rest)
+                                                 sql))))
+    query-dict))
+
+(defn- expand-params-native [{:keys [driver], :as query-dict} params]
+  (if-not (driver/driver-supports? driver :native-parameters)
+    query-dict
+    (let [report-timezone (get-in query-dict [:settings :report-timezone])
+          params          (flatten (map (partial expand-date-range-param report-timezone) params))]
+      (-> (substitute-all-params query-dict params)
+          remove-incomplete-clauses
+          process-multi-clauses
+          process-single-clauses))))
+
+
+;;; +-------------------------------------------------------------------------------------------------------+
+;;; |                                           PUBLIC API                                             |
```

Comment:
If we're going to use boxes like these to separate sections of code we might as well take an extra second to do it right. The `|` at the end of the second line is misaligned here.